controllers/resources/manager.go

```diff
@@ -136,55 +142,87 @@ func (m *manager) GenerateRemediationCRBase(gvk schema.GroupVersionKind) *unstru
 	// check if CR already exists
 	if err := m.Get(m.ctx, client.ObjectKeyFromObject(remediationCR), remediationCR); err == nil {
 		if !isOwner(remediationCR, nhc) {
 			m.log.Info("external remediation CR already exists, but it's not owned by us", "CR name", remediationCR.GetName(), "kind", remediationCR.GetKind(), "namespace", remediationCR.GetNamespace(), "owners", remediationCR.GetOwnerReferences())
-			return false, RemediationCRNotOwned{msg: "CR exists but isn't owned by current NHC"}
+			return false, nil, RemediationCRNotOwned{msg: "CR exists but isn't owned by current NHC"}
 		}
 		m.log.Info("external remediation CR already exists", "CR name", remediationCR.GetName(), "kind", remediationCR.GetKind(), "namespace", remediationCR.GetNamespace())
-		return false, nil
+		remediationCrs, err := m.ListRemediationCRs(nhc, func(cr unstructured.Unstructured) bool {
+			return cr.GetName() == remediationCR.GetName()
+		})
+		if err != nil {
+			m.log.Error(err, "couldn't fetch remediations for node", "node name", remediationCR.GetName())
+			return false, nil, err
+		}
+
+		duration, err := m.leaseManager.ManageLease(m.ctx, remediationCR, nhc, remediationCrs)
+		return false, &duration, err
 	} else if !apierrors.IsNotFound(err) {
 		m.log.Error(err, "failed to check for existing external remediation object")
-		return false, err
+		return false, nil, err
 	}
 
-	// create CR
-	m.log.Info("Creating a remediation CR",
-		"CR name", remediationCR.GetName(),
-		"CR kind", remediationCR.GetKind(),
-		"namespace", remediationCR.GetNamespace())
+	m.log.Info("Attempting to obtain Node Lease",
+		"Node name", remediationCR.GetName())
 
-	if err := m.Create(m.ctx, remediationCR); err != nil {
-		m.log.Error(err, "failed to create an external remediation object")
-		return false, err
+	isLeaseObtained, requeue, err := m.leaseManager.ObtainNodeLease(remediationCR, nhc)
+	if err != nil {
+		return false, requeue, err
 	}
-	return true, nil
+
+	if isLeaseObtained {
+		// create CR
+		m.log.Info("Creating a remediation CR",
+			"CR name", remediationCR.GetName(),
+			"CR kind", remediationCR.GetKind(),
+			"namespace", remediationCR.GetNamespace())
+
+		if err := m.Create(m.ctx, remediationCR); err != nil {
+			m.log.Error(err, "failed to create an external remediation object")
+			return false, nil, err
+		}
+	}
+
+	return isLeaseObtained, requeue, nil
+
 }
 
-func (m *manager) DeleteRemediationCR(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, error) {
+func (m *manager) DeleteRemediationCR(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck, remediationCrs []unstructured.Unstructured) (isDeleted bool, leaseRequeueIn *time.Duration, errResult error) {
+
+	defer func() {
+		if isDeleted {
+			//make sure lease will be deleted
+			remediationCrs = nil
```

Comment:
this is hard to understand... why do we set remediationCrs to nil here, and wy do we need them at all? :thinking:

Reply:
We fetch the remediation before the remediation is deleted, so `remediationCrs` shows a state which is no longer correct and will cause ManageLease to produce wrong output.